refactor(semantic): correct scope in CatchClause

Dunqing opened the pull request:
close: #4186

CatchClause has two scopes. The first one is `CatchClause`, which will add a `CatchParameter` to it. The second one is `Block`, which will add binding that declares in the current block scope.

The spec has a syntax error about `CatchParameter`
- It is a Syntax Error if any element of the BoundNames of CatchParameter also occurs in the LexicallyDeclaredNames of Block.

Dunqing commented:
* **#4192** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4192?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -321,7 +321,17 @@ impl<'a> SemanticBuilder<'a> {
+        let symbol_id = self.scope.get_binding(scope_id, name).or_else(|| {
+            self.scope.get_parent_id(scope_id).and_then(|pid| {
+                let flags = self.scope.get_flags(pid);
+                // It is a Syntax Error if any element of the BoundNames of CatchParameter also occurs in the LexicallyDeclaredNames of Block.
```
I don't think this check is in the right place 🤔

Dunqing replied:
I tried another way. Can you take a look at https://github.com/oxc-project/oxc/pull/4192/commits/3ed49909c2a226961b92601e11cee42108c4973a?

Dunqing committed:
refactor(semantic): correct scope in CatchClause (#4192)

close: #4186

CatchClause has two scopes. The first one is `CatchClause`, which will add a `CatchParameter` to it. The second one is `Block`, which will add binding that declares in the current block scope.

The spec has a syntax error about `CatchParameter`
- It is a Syntax Error if any element of the BoundNames of CatchParameter also occurs in the LexicallyDeclaredNames of Block.

overlookmotel commented on crates/oxc_ast/src/ast/js.rs:
```diff
@@ -1300,7 +1300,7 @@ pub struct TryStatement<'a> {
 }
 
 #[visited_node]
-#[scope(if(self.param.is_some()))]
```
@Dunqing What was reason for removing `if(self.param.is_some())`? If no catch param, we don't need a scope, right?

I think it's arguable that it would be simpler if we didn't have any conditional scopes, but if we do that, probably we should do it consistently.

Dunqing replied:
Yes, just to be consistent with the spec

overlookmotel replied:
OK, so is plan to make all scopes unconditional?

overlookmotel replied:
Personally, I don't think we *need* to adhere to the spec. The spec defines what *observable behavior* must be, but implementations are free to achieve that through any means they choose. V8, for example, diverges from the example algorithms in the spec very heavily for purpose of better performance.

Dunqing replied:
I agree. I have found that there is a performance impact of `boa` reporting errors according to the specification

https://github.com/boa-dev/boa/blob/5b1621abd2b702b99bd6b4491973095f9b27c362/core/parser/src/parser/statement/try_stm/catch.rs#L92-L97

overlookmotel replied:
Thanks for coming back. OK. I would suggest reverting this to `#[scope(if(self.param.is_some()))]` for now then. If we keep scopes as conditional, it's quite easy to make them all unconditional later. But it's harder to go in other direction - you have to go through all the scopes and reason about which *could* be conditional.

Personally I think the most important thing either way is to be consistent, as it's one less oddity to reason about.

But that's just my opinion. You can disagree, and I won't argue further.

Dunqing replied:
Ok. I will send a PR later
